airflow/jobs/scheduler_job_runner.py

```diff
@@ -986,11 +1023,23 @@ def _run_scheduler_loop(self) -> None:
                     self.processor_agent.wait_until_finished()
 
                 with create_session() as session:
+                    # This will schedule for as many executors as possible.
                     num_queued_tis = self._do_scheduling(session)
 
-                    self.job.executor.heartbeat()
+                    # Heartbeat all executors, even if they're not receiving new tasks this loop. It will be
+                    # either a no-op, or they will check-in on currently running tasks and send out new
+                    # events to be processed below.
+                    for executor in self.job.executors:
+                        executor.heartbeat()
+
+                    # TODO: I'm not sure why we expunge here. Do we need to do it after each individual
```

Comment:
Todo left intentionally. I would like community feedback on this one!

Reply:
Seemingly random expunges are usually for HA (running multiple scheduler processes). This seems to be the case; the line was first introduced in 73b9163a8f55ce3d5bf6aec0a558952c27dd1b55. So this is fine—*maybe* we can do this before heartbeating to avoid some confusion? But that might mess up some edge cases a bit since expunging might take some time (???) not exactly sure.

Reply:
I believe (@ashb?) it's been added to isolate scheduling/enqueuing tasls and acompanied DB operations from executors executing tasks. Simply it might happen that parallel scheduler will make some changes to objects (DAGs) in parallel when scheduling things and the idea was - I think - to make sure that executors will act on the latest version of that data (after expunge_all, all the executor operations will have to read from DB to get the objects).

I think we should not move it before heartbeating - because heartbeating does most ol the work of executing tasks, so that would change the behaviour a lot.

Reply:
> I think we should not move it before heartbeating

That's fair, and I agree, my real question is: can I do it _once_ after all the executors have been heartbeated (what I'm doing now), or does it need to be done after heartbeating each individual executor. Given your context, I'm more confident that we can do it once after all executors have been heartbeated. But correct me if I'm wrong there!

Reply:
I think you are right :)